### PR TITLE
Fieldset support in Attributes/JSON adapters

### PR DESCRIPTION
**Pull request by kosmatov**
Adds the ability to specify the returned relation fields in Attributes and JSON adapters

### CODE REVIEW

**Comment by bf4** [comment]
I like this.. will need to review a little more closely

File: lib/active_model_serializers/adapter/base.rb [hunk 1]
```diff
@@ -78,6 +83,32 @@ def serialization_options(options)
       def root
         serializer.json_key.to_sym if serializer.json_key
       end
+
+      def fields_to_hash(fields)
+        case fields
+        when Hash
+          fields
+        when Array
```
Comment (BenMorganIO) [inline lib/active_model_serializers/adapter/base.rb]: To follow the ruby philosophy, I think it would be of a greater benefit if we used `respond_to?`. This way you can send in object that implement `Hash` or `Array`, but aren't necessarily those objects themselves.

File: lib/active_model_serializers/adapter/base.rb [hunk 2]
```diff
@@ -46,6 +46,7 @@ def self.fragment_cache(cached_hash, non_cached_hash)
 
       def initialize(serializer, options = {})
         @serializer = serializer
+        @fieldset = options[:fieldset] || ActiveModel::Serializer::Fieldset.new(fields_to_hash(options.delete(:fields)))
```
Comment (bf4) [inline lib/active_model_serializers/adapter/base.rb]: in jsonapi was

```
@fieldset = options[:fieldset] || ActiveModel::Serializer::Fieldset.new(options.delete(:fields))
```

why was `fields_to_hash` introduced here?

File: lib/active_model_serializers/adapter/json.rb [hunk 3]
```diff
@@ -3,6 +3,7 @@ module Adapter
     class Json < Base
       def serializable_hash(options = nil)
         options = serialization_options(options)
+        instance_options[:fieldset] = fieldset
```
Comment (bf4) [inline lib/active_model_serializers/adapter/json.rb]: why is this here and not in the base adapter? 

why in the attributes adapter is it `options[:fields] ||= fields || instance_options[:fields]` both adapters do basically the same thing, no?

File: lib/active_model_serializers/adapter/attributes.rb [hunk 4]
```diff
@@ -2,8 +2,8 @@ module ActiveModelSerializers
   module Adapter
     class Attributes < Base
       def serializable_hash(options = nil)
-        options = serialization_options(options)
-        options[:fields] ||= instance_options[:fields]
+        options ||= serialization_options(options)
```
Comment (bf4) [inline lib/active_model_serializers/adapter/attributes.rb]: why was `options =` changed to `options ||=`?

File: lib/active_model/serializer.rb [hunk 5]
```diff
@@ -379,6 +379,7 @@ def associations(include_directive = ActiveModelSerializers.default_include_dire
     def serializable_hash(adapter_options = nil, options = {}, adapter_instance = self.class.serialization_adapter_instance)
       adapter_options ||= {}
       options[:include_directive] ||= ActiveModel::Serializer.include_directive_from_options(adapter_options)
+      options[:fields] ||= adapter_options[:fieldset] && adapter_options[:fieldset].fields_for(json_key)
```
Comment (bf4) [inline lib/active_model/serializer.rb]: This seems to duplicate logic in the adapter.

```
def fields
   fieldset && fieldset.fields_for(serializer.json_key)
end
```

which should it be?

Is the problem here that the adapter does `options.delete(:fields)`?

**Comment by wasifhossain** [comment]
closing this as https://github.com/rails-api/active_model_serializers/pull/2223 is merged
